src/ape/exceptions.py

```diff
@@ -113,9 +114,82 @@ def __init__(
         self.txn = txn
         self.trace = trace
         self.contract_address = contract_address
+        self.source_traceback = None
         ex_message = f"({code}) {message}" if code else message
         super().__init__(ex_message)
 
+        # Attempt to set stacktrace.
+        if not txn:
+            return
+
+        receipt = txn.receipt
+        if not receipt:
+            return
+
+        try:
+            ape_traceback = receipt.source_traceback
+        except (ApeException, NotImplementedError):
+            return
+
+        if not ape_traceback or not len(ape_traceback):
+            return
+
+        self.source_traceback = ape_traceback
+
+        # Manipulate python traceback to show lines from contract.
+        # Help received from Jinja lib:
+        #  https://github.com/pallets/jinja/blob/main/src/jinja2/debug.py#L142
+        _, exc_value, tb = sys.exc_info()
+        depth = None
+        idx = len(ape_traceback) - 1
+        stack = []
+        project_path = txn.project_manager.path.as_posix()
+        while tb is not None:
+            if not tb.tb_frame.f_code.co_filename.startswith(project_path):
+                # Ignore frames outside the project.
+                # This allows both contract code an scripts to appear.
+                tb = tb.tb_next
+                continue
+
+            stack.append(tb)
+            tb = tb.tb_next
+
+        while (depth is None or depth > 1) and idx >= 0:
+            exec_item = ape_traceback[idx]
+            if depth is not None and exec_item.depth >= depth:
+                # Wait for decreasing depth.
+                continue
+
+            depth = exec_item.depth
+            lineno = exec_item.begin_lineno
+            filename = exec_item.source_path.as_posix()
+
+            # Raise an exception at the correct line number.
+            py_code: CodeType = compile(
+                "\n" * (lineno - 1) + "raise __ape_exception__", filename, "exec"
+            )
+            py_code = py_code.replace(co_name=exec_item.function.signature)
+
+            # Execute the new code to get a new (fake) tb with contract source info.
+            try:
+                exec(py_code, {"__ape_exception__": self}, {})
+            except BaseException:
+                fake_tb = sys.exc_info()[2].tb_next  # type: ignore
+                if isinstance(fake_tb, TracebackType):
+                    stack.append(fake_tb)
+
+            idx -= 1
+
+        if not stack:
+            return
+
+        tb_next = None
+        for tb in stack:
+            tb.tb_next = tb_next
+            tb_next = tb
+
+        self.__traceback__ = stack[-1]
```

Comment:
Sorta wish this code was a utility function

Reply:
i can move it to a method in this same module but cant put it in utils because of circular deps

Reply:
That's good enough